Check XFB indexed bindings during bufferData calls

lexaknyazev opened the pull request:
#### 4c366686ccd2bb9b9664f4e0a52c241b4ac9bcb8
<pre>
Check XFB indexed bindings during bufferData calls
<a href="https://bugs.webkit.org/show_bug.cgi?id=223358">https://bugs.webkit.org/show_bug.cgi?id=223358</a>

Reviewed by Kimmo Kinnunen.

A buffer which is simultaneously bound to an indexed
TRANSFORM_FEEDBACK_BUFFER binding point in the currently bound
transform feedback object and any other binding point in the WebGL API,
except generic TRANSFORM_FEEDBACK_BUFFER binding point, cannot be used.

Fixes conformance2/transform_feedback/simultaneous_binding.html

Drive-by:
* Skip extra error generation in copyBufferSubData and getBufferSubData.
* Remove deleted buffer from indexed uniform bindings.

* LayoutTests/TestExpectations:
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::validateBufferDataTarget):
(WebCore::WebGL2RenderingContext::copyBufferSubData):
(WebCore::WebGL2RenderingContext::getBufferSubData):
(WebCore::WebGL2RenderingContext::uncacheDeletedBuffer):
* Source/WebCore/html/canvas/WebGLTransformFeedback.h:
(WebCore::WebGLTransformFeedback::hasBoundIndexedTransformFeedbackBuffer):
* Source/WebCore/html/canvas/WebGLVertexArrayObjectBase.h:
(WebCore::WebGLVertexArrayObjectBase::hasArrayBuffer):

Canonical link: <a href="https://commits.webkit.org/253210@main">https://commits.webkit.org/253210@main</a>
</pre>

djg commented on Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
```diff
@@ -591,11 +605,12 @@ void WebGL2RenderingContext::copyBufferSubData(GCGLenum readTarget, GCGLenum wri
 
     RefPtr<WebGLBuffer> readBuffer = validateBufferDataParameters("copyBufferSubData", readTarget, GraphicsContextGL::STATIC_DRAW);
+    if (!readBuffer)
+        return;
```
Is the "Invalid readTarget" synthetic error not required here?

lexaknyazev replied:
`validateBufferDataParameters` or its callees generate a synthetic error before returning `nullptr`.

djg commented on Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
```diff
@@ -591,11 +605,12 @@ void WebGL2RenderingContext::copyBufferSubData(GCGLenum readTarget, GCGLenum wri
     RefPtr<WebGLBuffer> writeBuffer = validateBufferDataParameters("copyBufferSubData", writeTarget, GraphicsContextGL::STATIC_DRAW);
-    if (!readBuffer || !writeBuffer) {
-        synthesizeGLError(GraphicsContextGL::INVALID_OPERATION, "copyBufferSubData", "Invalid readTarget or writeTarget");
+    if (!writeBuffer)
```
Is the "Invalid writeTarget" synthetic error not required here?

lexaknyazev replied:
`validateBufferDataParameters` or its callees generate a synthetic error before returning `nullptr`.

kkinnunen-apple reviewed (approved):
Nice!

lexaknyazev committed:
Check XFB indexed bindings during bufferData calls
https://bugs.webkit.org/show_bug.cgi?id=223358

Reviewed by Kimmo Kinnunen.

A buffer which is simultaneously bound to an indexed
TRANSFORM_FEEDBACK_BUFFER binding point in the currently bound
transform feedback object and any other binding point in the WebGL API,
except generic TRANSFORM_FEEDBACK_BUFFER binding point, cannot be used.

Fixes conformance2/transform_feedback/simultaneous_binding.html

Drive-by:
* Skip extra error generation in copyBufferSubData and getBufferSubData.
* Remove deleted buffer from indexed uniform bindings.

* LayoutTests/TestExpectations:
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::validateBufferDataTarget):
(WebCore::WebGL2RenderingContext::copyBufferSubData):
(WebCore::WebGL2RenderingContext::getBufferSubData):
(WebCore::WebGL2RenderingContext::uncacheDeletedBuffer):
* Source/WebCore/html/canvas/WebGLTransformFeedback.h:
(WebCore::WebGLTransformFeedback::hasBoundIndexedTransformFeedbackBuffer):
* Source/WebCore/html/canvas/WebGLVertexArrayObjectBase.h:
(WebCore::WebGLVertexArrayObjectBase::hasArrayBuffer):

Canonical link: https://commits.webkit.org/253210@main
